### PR TITLE
Bugfix/prevent site collection (closes #6785)

### LINKED ISSUE #6785: All of site-packages is vendored when collecting data or binaries from single-file modules (like `appdirs`)
## Description of the issue

When packaging an application that depends on the [`appdirs`](https://pypi.org/project/appdirs/) library, if `--collect-data` or `--collect-binaries` is included in the invocation, the resulting distribution has a complete copy of the host's `site-packages`.

I'm not sure if this is a bug with PyInstaller or a problem with this particular package, but I'm filing here in the hopes of clarifying that one way or the other.

### Context information (for bug reports)

* Output of `pyinstaller --version`: 5.0.1
* Version of Python: 3.8
* Platform: Ubuntu 20.04 (also observed on Windows 10)
* How you installed Python: pyenv / official installer

* follow *all* the instructions in our "If Things Go Wrong" Guide
  (https://github.com/pyinstaller/pyinstaller/wiki/If-Things-Go-Wrong) and

### Make sure [everything is packaged correctly](https://github.com/pyinstaller/pyinstaller/wiki/How-to-Report-Bugs#make-sure-everything-is-packaged-correctly)

  * [ ] start with clean installation
  * [ ] use the latest development version
  * [ ] Run your frozen program **from a command window (shell)** — instead of double-clicking on it
  * [ ] Package your program in **--onedir mode**
  * [ ] Package **without UPX**, say: use the option `--noupx` or set `upx=False` in your .spec-file
  * [ ] Repackage you application in **verbose/debug mode**. For this, pass the option `--debug` to `pyi-makespec` or `pyinstaller` or use `EXE(..., debug=1, ...)` in your .spec file.

### Good case
```
$ echo 'print("test program")' > test.py
$ python3 -m PyInstaller test.py
...
$ tree -L 2 dist/test/
dist/test/
├── base_library.zip
├── lib-dynload
│   ├── _asyncio.cpython-38-x86_64-linux-gnu.so
│   ├── _bz2.cpython-38-x86_64-linux-gnu.so
│   ├── _codecs_cn.cpython-38-x86_64-linux-gnu.so
│   ├── _codecs_hk.cpython-38-x86_64-linux-gnu.so
│   ├── _codecs_iso2022.cpython-38-x86_64-linux-gnu.so
│   ├── _codecs_jp.cpython-38-x86_64-linux-gnu.so
│   ├── _codecs_kr.cpython-38-x86_64-linux-gnu.so
│   ├── _codecs_tw.cpython-38-x86_64-linux-gnu.so
│   ├── _contextvars.cpython-38-x86_64-linux-gnu.so
│   ├── _ctypes.cpython-38-x86_64-linux-gnu.so
│   ├── _decimal.cpython-38-x86_64-linux-gnu.so
│   ├── _hashlib.cpython-38-x86_64-linux-gnu.so
│   ├── _lzma.cpython-38-x86_64-linux-gnu.so
│   ├── _multibytecodec.cpython-38-x86_64-linux-gnu.so
│   ├── _multiprocessing.cpython-38-x86_64-linux-gnu.so
│   ├── _opcode.cpython-38-x86_64-linux-gnu.so
│   ├── _posixshmem.cpython-38-x86_64-linux-gnu.so
│   ├── _queue.cpython-38-x86_64-linux-gnu.so
│   ├── _ssl.cpython-38-x86_64-linux-gnu.so
│   ├── mmap.cpython-38-x86_64-linux-gnu.so
│   ├── readline.cpython-38-x86_64-linux-gnu.so
│   ├── resource.cpython-38-x86_64-linux-gnu.so
│   └── termios.cpython-38-x86_64-linux-gnu.so
├── libbz2.so.1.0
├── libcrypto.so.1.1
├── libexpat.so.1
├── libffi.so.7
├── liblzma.so.5
├── libmpdec.so.2
├── libpython3.8.so.1.0
├── libreadline.so.8
├── libssl.so.1.1
├── libtinfo.so.6
├── libz.so.1
└── test

1 directory, 36 files
```

### Bad case
```
$ echo 'print("test program")' > test.py
$ python3 -m PyInstaller --collect-data appdirs test.py
...
$ tree -L 2 dist/test/
dist/test/
├── base_library.zip
├── lib-dynload
│   ├── _asyncio.cpython-38-x86_64-linux-gnu.so
│   ├── _bz2.cpython-38-x86_64-linux-gnu.so
│   ├── _codecs_cn.cpython-38-x86_64-linux-gnu.so
│   ├── _codecs_hk.cpython-38-x86_64-linux-gnu.so
│   ├── _codecs_iso2022.cpython-38-x86_64-linux-gnu.so
│   ├── _codecs_jp.cpython-38-x86_64-linux-gnu.so
│   ├── _codecs_kr.cpython-38-x86_64-linux-gnu.so
│   ├── _codecs_tw.cpython-38-x86_64-linux-gnu.so
│   ├── _contextvars.cpython-38-x86_64-linux-gnu.so
│   ├── _ctypes.cpython-38-x86_64-linux-gnu.so
│   ├── _decimal.cpython-38-x86_64-linux-gnu.so
│   ├── _hashlib.cpython-38-x86_64-linux-gnu.so
│   ├── _lzma.cpython-38-x86_64-linux-gnu.so
│   ├── _multibytecodec.cpython-38-x86_64-linux-gnu.so
│   ├── _multiprocessing.cpython-38-x86_64-linux-gnu.so
│   ├── _opcode.cpython-38-x86_64-linux-gnu.so
│   ├── _posixshmem.cpython-38-x86_64-linux-gnu.so
│   ├── _queue.cpython-38-x86_64-linux-gnu.so
│   ├── _ssl.cpython-38-x86_64-linux-gnu.so
│   ├── mmap.cpython-38-x86_64-linux-gnu.so
│   ├── readline.cpython-38-x86_64-linux-gnu.so
│   ├── resource.cpython-38-x86_64-linux-gnu.so
│   └── termios.cpython-38-x86_64-linux-gnu.so
├── libbz2.so.1.0
├── libcrypto.so.1.1
├── libexpat.so.1
├── libffi.so.7
├── liblzma.so.5
├── libmpdec.so.2
├── libpython3.8.so.1.0
├── libreadline.so.8
├── libssl.so.1.1
├── libtinfo.so.6
├── libz.so.1
├── site-packages
│   ├── PyInstaller
│   ├── _pyinstaller_hooks_contrib
│   ├── altgraph-0.17.2.dist-info
│   ├── appdirs-1.4.4.dist-info
│   ├── pip-20.0.2.dist-info
│   ├── pkg_resources-0.0.0.dist-info
│   ├── pyinstaller-5.0.1.dist-info
│   ├── pyinstaller_hooks_contrib-2022.4.dist-info
│   ├── setuptools
│   └── setuptools-44.0.0.dist-info
└── test

12 directories, 36 files
```

### non-packaging reproduction

Output is truncated, but running `collect_data_files()` on the package returns a list of the entire site.

```
$ python3  # in an environment with appdirs installed
Python 3.8.10 (default, Mar 15 2022, 12:22:08)
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from PyInstaller.utils.hooks import collect_data_files
>>> collect_data_files("appdirs")
[('/home/snoopjedi/scratch/pyinstaller_repro/venv/lib/python3.8/site-packages/setuptools-44.0.0.dist-info/AUTHORS.txt', 'site-packages/setuptools-44.0.0.dist-info'),  ... ]
```

### PR DESCRIPTION
**Pull request by SnoopJ**
This PR adds an `is_package()` check to the data/binary collection routines to prevent duplicating the entire site of the host interpreter when called for single-file libraries, see #6785

### CODE REVIEW

File: news/6789.bugfix.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Prevent collection of an entire Python site for single-file modules
```
Comment (rokm) [inline news/6789.bugfix.rst]: Prevent collection of an entire Python site when using ``collect_data_files`` against single-file modules.

Comment (bwoodsend) [reply news/6789.bugfix.rst]: Or you can really overcompensate and make them cross references ``:cmdoption:`--collect-data` `` or ``:func:`~PyInstaller.utils.hooks.collect_data_files` ``.

**Review by rokm — APPROVED** [review]

**Comment by SnoopJ** [comment]
That CI failure (3x `OSError(24, 'Too many open files')`) left me scratching my head. Can one of y'all have a look?

**Comment by bwoodsend** [comment]
Oh, don't worry about that one. It's been floating around for a while.

**Comment by bwoodsend** [comment]
Thanks for mopping that up for me.

**Pull request merged**